src/core/index.tsx

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+import React from 'react'
+
+import {
+  TestHookProps,
+  NativeRendererOptions,
+  NativeRendererReturn,
+  ResultContainerReturn,
+  RenderHookOptions,
+  RenderResult,
+  ServerRendererReturn,
+  ServerRendererOptions,
+  DomRendererReturn,
+  RenderHookReturn,
+  ServerRenderHook
+} from '../types'
+
+import asyncUtils from './asyncUtils'
+import { cleanup, addCleanup, removeCleanup } from './cleanup'
+
+function resultContainer<TValue>(): ResultContainerReturn<TValue> {
+  const results: Array<{ value?: TValue; error?: Error }> = []
+  const resolvers: Array<() => void> = []
+
+  const result: RenderResult<TValue> = {
+    get all() {
+      return results.map(({ value, error }) => error ?? value)
+    },
+    get current() {
+      const { value, error } = results[results.length - 1]
+      if (error) {
+        throw error
+      }
+      return value as TValue
+    },
+    get error() {
+      const { error } = results[results.length - 1]
+      return error
+    }
+  }
+
+  const updateResult = (value?: TValue, error?: Error) => {
+    results.push({ value, error })
+    resolvers.splice(0, resolvers.length).forEach((resolve) => resolve())
+  }
+
+  return {
+    result,
+    addResolver: (resolver: () => void) => {
+      resolvers.push(resolver)
+    },
+    setValue: (value: TValue) => updateResult(value),
+    setError: (error: Error) => updateResult(undefined, error)
+  }
+}
+
+// typed this way in relation to this https://github.com/DefinitelyTyped/DefinitelyTyped/issues/44572#issuecomment-625878049
+function defaultWrapper({ children }: { children?: React.ReactNode }) {
+  return (children as unknown) as JSX.Element
+}
+
+function createRenderHook(
+  createRenderer: <TProps, TResult>(
+    testProps: Omit<TestHookProps<TProps, TResult>, 'hookProps'>,
+    opts: ServerRendererOptions<TProps>
+  ) => ServerRendererReturn<TProps>
+): <TProps, TResult>(
+  callback: (props: TProps) => TResult,
+  opts?: RenderHookOptions<TProps>
+) => ServerRenderHook<TProps, TResult>
+function createRenderHook(
+  createRenderer: <TProps, TResult>(
+    testProps: Omit<TestHookProps<TProps, TResult>, 'hookProps'>,
+    opts: NativeRendererOptions<TProps>
+  ) => NativeRendererReturn<TProps> | DomRendererReturn<TProps>
+): <TProps, TResult>(
+  callback: (props: TProps) => TResult,
+  opts?: RenderHookOptions<TProps>
+) => RenderHookReturn<TProps, TResult>
+function createRenderHook(
```

Comment:
I've got some ideas to reduce the noise here and make `createRenderHook` agnostic of any reference to the type of renderer.  It's a bit difficult to explain, so I might put up a PR to your branch with what I'm thinking tomorrow.

Reply:
That would be great, my main challenge was have TS understand when hydrate would exist and when it wouldn't, hence the overload.

Reply:
The idea I had would infer extra functions from the result of `createRenderer` and the resulting type _just works™️_.  The function signature might get a bit crazy, but the usage won't change.  We'll see how it goes 😅